clang/lib/Sema/SemaDeclAttr.cpp

```diff
@@ -10,6 +10,7 @@
 //
 //===----------------------------------------------------------------------===//
 
+#include "../Basic/Targets/NVPTX.h"
```

Comment:
This header is not part of `clangBasic`'s interface, but rather its implementation (`lib/Basic` rather than `include/clang/Basic`).
Sema can't depend on it - if you need to use its APIs from outside clangBasic they should be moved to a public header.

The bazel build shows the problem: https://buildkite.com/llvm-project/upstream-bazel/builds/75928#018ad568-2f7c-4dda-ae90-3b4d787caad7. Breaking bazel is not itself reason to revert, but here it's flagging a real problem that CMake doesn't catch.

I've reverted as 0afbcb20fd908f8bf9073697423da097be7db592 - sorry to do this so abruptly, but I can't fix this myself & such problems block downstream use of LLVM.

Reply:
@sam-mccall, apologies for introducing the bug and thank you for drawing my attention to it.

I've got the fix for the problem:
```diff
diff --git a/clang/lib/Sema/SemaDeclAttr.cpp b/clang/lib/Sema/SemaDeclAttr.cpp
index 10d1c910d9cd..3b87300e24bc 100644
--- a/clang/lib/Sema/SemaDeclAttr.cpp
+++ b/clang/lib/Sema/SemaDeclAttr.cpp
@@ -10,7 +10,6 @@
 //
 //===----------------------------------------------------------------------===//

-#include "../Basic/Targets/NVPTX.h"
 #include "clang/AST/ASTConsumer.h"
 #include "clang/AST/ASTContext.h"
 #include "clang/AST/ASTMutationListener.h"
@@ -5612,7 +5611,8 @@ bool Sema::CheckRegparmAttr(const ParsedAttr &AL, unsigned &numParams) {
 static CudaArch getCudaArch(const TargetInfo &TI) {
   if (!TI.getTriple().isNVPTX())
     llvm_unreachable("getCudaArch is only valid for NVPTX triple");
-  return static_cast<const targets::NVPTXTargetInfo *>(&TI)->getGPU();
+  auto &TO = TI.getTargetOpts();
+  return StringToCudaArch(TO.CPU);
 }

 // Checks whether an argument of launch_bounds attribute is
```

Would you be so king and point me to the process for "reverting the revert" and folding the fix into the original patch?

Reply:
@jchlanda ah, that's simpler than I expected! Wish I'd found that before reverting...

That looks good to me, I think rather than a new review you can just `git revert 0afbcb2`, make the changes you described above, `git amend -a` and change the description to "Reland [NVPTX] ...", run the tests and push.

Or if you prefer, send it as a new PR, happy to approve it.